fw/sock_clnt.c

```diff
@@ -627,21 +626,71 @@ tfw_sock_clnt_cfgend(void)
 static int
 tfw_sock_clnt_start(void)
 {
-	int r;
+	int r = 0;
 	TfwListenSock *ls;
+	TfwListenSock *ls_reconf;
+	TfwListenSock *tmp;
 
-	if (tfw_runstate_is_reconfig())
-		return 0;
+	tfw_classifier_cleanup_inport();
+	list_for_each_entry(ls_reconf, &tfw_listen_socks_reconf, list) {
+		bool found = false;
+
+		list_for_each_entry(ls, &tfw_listen_socks, list) {
+			if (tfw_addr_eq(&ls_reconf->addr, &ls->addr)) {
+				found = true;
+				break;
+			}
+		}
 
-	list_for_each_entry(ls, &tfw_listen_socks, list) {
-		if ((r = tfw_listen_sock_start(ls))) {
-			T_ERR_ADDR("can't start listening on", &ls->addr,
-				   TFW_WITH_PORT);
-			goto err;
+		if (!found) {
+			ls = kzalloc(sizeof(*ls), GFP_KERNEL);
+			if (!ls) {
+				r = -ENOMEM;
+				goto err;
+			}
+
+			ss_proto_init(&ls->proto, ls_reconf->proto.hooks, ls_reconf->proto.type);
+
+			list_add(&ls->list, &tfw_listen_socks);
+			ls->addr = ls_reconf->addr;
+
+			if ((r = tfw_listen_sock_start(ls))) {
+				T_ERR_ADDR("can't start listening on", &ls_reconf->addr,
+					   TFW_WITH_PORT);
+				goto err;
+			}
 		}
+
 	}
 
-	return 0;
+	list_for_each_entry_safe(ls, tmp, &tfw_listen_socks, list) {
+		bool found = false;
+
+		list_for_each_entry(ls_reconf, &tfw_listen_socks_reconf, list) {
+			if (tfw_addr_eq(&ls_reconf->addr, &ls->addr)) {
+				found = true;
+				break;
+			}
+		}
+
+		if (!found) {
+			list_del(&ls->list);
+			if (!ls->sk)
+				continue;
+			ss_release(ls->sk);
```

Comment:
If we close the port, then we also should remove it from the classifier. Since classifier is used in runtime, I think it makes sense to create a new ports array and just substitute them in RCU fashion